include/zephyr/drivers/rtcc.h

```diff
@@ -0,0 +1,241 @@
+ * @note This handler is likely to be invoked from within an ISR context
+ * @param dev Device instance invoking the handler
+ * @param id Alarm id
+ * @param user_data Optional user_data provided with the alarm configuration
+ */
+typedef void (*rtcc_alarm_triggered_handler_t)(const struct device *dev, size_t id,
+					       void *user_data);
+
+/**
+ * @name RTCC alarm trigger mask
+ * @brief Mask for tm fields to trigger alarm
+ * @{
+ */
+#define RTCC_ALARM_MATCH_SECOND          BIT(0)
+#define RTCC_ALARM_MATCH_MINUTE          BIT(1)
+#define RTCC_ALARM_MATCH_HOUR            BIT(2)
+#define RTCC_ALARM_MATCH_DAY             BIT(3)
+#define RTCC_ALARM_MATCH_MONTH           BIT(4)
+#define RTCC_ALARM_MATCH_YEAR            BIT(5)
+/**
+ * @}
+ */
+
+/**
+ * @typedef rtcc_alarm_trigger_mask
+ * @brief Type storing bitmask of RTCC alarm triggers
+ */
+typedef uint8_t rtcc_alarm_match_mask_t;
+
+/**
+ * @brief RTCC alarm configuration
+ * @param time Date and time to match
+ * @param mask Parameters of the date and time to match
+ * @param handler Handler invoked once the alarm is triggered
+ * @param user_data Optional user data which passed to the handler when invoked
+ * @note Handler and user data will be lost of device resets before alarm is triggered
+struct rtcc_alarm_config {
+	struct tm time;
+	rtcc_alarm_match_mask_t mask;
+	rtcc_alarm_triggered_handler_t handler;
+	void *user_data;
+};
+
+/**
+ * @typedef rtcc_time_set_t
+ * @brief API for setting RTCC date and time
+ * @param dev Device instance
+ * @param time The date and time to set
+ */
+typedef int (*rtcc_time_set_t)(const struct device *dev, const struct tm *time);
+
+/**
+ * @typedef rtcc_time_get_t
+ * @brief API for getting RTCC date and time
+ * @param dev Device instance
+ * @param time Destination for the current date and time
+ */
+typedef int (*rtcc_time_get_t)(const struct device *dev, struct tm *time);
+
+/**
+ * @typedef rtcc_alarm_start_t
+ * @brief API for configuring and setting an RTCC alarm
+ * @param dev Device instance
+ * @param id Id of the alarm to configure
+ * @param config The alarm configuration to use
+ */
+typedef int (*rtcc_alarm_set_t)(const struct device *dev, size_t id,
+				const struct rtcc_alarm_config *config);
+
+/**
+ * @typedef rtcc_alarm_cancel_t
+ * @brief API for cancelling an RTCC alarm
+ * @param dev Device instance
+ * @param id Id of the alarm to cancel
+ */
+typedef int (*rtcc_alarm_cancel_t)(const struct device *dev, size_t id);
+
+/**
+ * @typedef rtcc_timestamp_get_t
+ * @brief API for getting timestamp
+ * @param dev Device instance
+ * @param id Id of the alarm to cancel
+ */
+typedef int (*rtcc_timestamp_get_t)(const struct device *dev, size_t id, struct tm *time);
+
+/**
+ * @brief RTCC driver API
+ */
+__subsystem struct rtcc_driver_api {
+	rtcc_time_set_t time_set;
+	rtcc_time_get_t time_get;
+	rtcc_alarm_start_t alarm_start;
+	rtcc_alarm_cancel_t alarm_cancel;
+	rtcc_timestamp_get_t timestamp_get;
```

Comment:
Hi, I am trying to develop an individual RTC device for MC146818B RTC, available on many x86 boards. 
I have checked these header files and rtcc.h will be very useful. But this could have two more APIs which will support READ/WRITE of CMOS battery backed SRAM (256 bytes in my case). 
APIs similar to 

uint8_t (*read_offset)(const struct device *dev, struct rtc_address *addr) &
int (*write_offset)(const struct device *dev, struct rtc_address *addr, uint8_t value) 
and supporting data types,

struct rtc_address {
	/* RAM bank */
	enum rtc_ram_bank rtc_bank;
	/* offset */
	uint8_t offset;
};
 
enum rtc_ram_bank {
	RTC_STD_RAM,
	RTC_EXT_RAM,
};
This enum helps to identify between 2 blocks of SRAM.

This memory could be used by BIOS or for any other services.

Reply:
The retention registers are not specific to an RTC module, some I2C attached sensors also allow for retained registers, and some MCUs have internal retention registers and retained SRAM.

We need a tailored API for reading/writing to SRAM, retention registers, and other forms of retained memory, like external EEPROMs.

This functionality is then added to the external RTC node in the devicetree as such:
```
&i2c2 {
    pinctrl-0 = <&i2c2_scl_ph4 &i2c2_sda_ph5>;
    pinctrl-names = "default";
    status = "okay";
    clock-frequency = <I2C_BITRATE_FAST>;

    /* Implements RTCC API */
    pca8565@2a {
        compatible = "pca8565";
        reg = <0x2a>;

        /* Implements said reg read/write API */
        backup_regs {
            compatible =  "pca8565_backup_regs";
        };	
    };
};
```

This issue https://github.com/zephyrproject-rtos/zephyr/issues/51298 has some more discussion around this. We still don't have an API for performing this, because it is actually quite complex to create a solution that can be used effectively in the devicetree and as a driver API :)

Reply:
It's quite possible that another API and device tree binding already exists that would address this. We don't need to have separate API just for reading and writing at an offset.

E.g. flash.sh

Reply:
> Hi, thank you suggestion. I found that "bbram.h" will serve the purpose for reading/writing to battery backed RAM.
> 
> Regarding the APIs in rtcc.h, could there be a possibility to add an API to get current alarm time like rtcc_get_alarm.

We can definitely can add that. Can you think of a use case for getting the timer configuration? The Raspberry Pi Pico SDK doesn't provide an `alarm_get` https://raspberrypi.github.io/pico-sdk-doxygen/group__hardware__rtc.html And what should happen if the timer is not configured, should we return an error code maybe?